### ISSUE TITLE
Node contraction may return incorrect results in cases where groups consist of only one node

### ISSUE THREAD

**Issue by mkoohafkan** [thomasp85/tidygraph#140]
Originally discovered by @mgastner:

> The following network only has two nodes (A and B) and one directed edge A -> B. I would have expected that the network contracted by node ID would simply return the original network, but, when plotting the network, the direction of the edge is reversed.
```r
library(ggraph)
library(tidygraph)
library(tidyverse)
demo_nodes <- tibble(node_id = c("B", "A"))
demo_edges <- tibble(from = "A", to = "B")
demo_netw <- tbl_graph(nodes = demo_nodes, edges = demo_edges)
demo_netw_contracted <-
  demo_netw %>%
  convert(to_contracted, node_id)
ggraph(demo_netw_contracted) +
  geom_edge_fan(aes(end_cap = label_rect(node2.node_id)), arrow = arrow()) +
  geom_node_label(aes(label = node_id))
```
Using `igraph::contract()` does what you expect and returns the original graph:
```r
library(igraph)
contract(as.igraph(demo_netw), 1:2)
```

Here's the culprit: https://github.com/thomasp85/tidygraph/blob/75c814ee9c45b2514396209939b7f06509a36b8a/R/morphers.R#L251

we'll reproduce the logic using this example:

```r
nodes <- as_tibble(demo_netw, active = 'nodes')
nodes <- group_by(nodes, node_id)
ind <- group_indices(nodes)
ind
## [1] 2 1
```
For whatever reason, node id "B" is assigned the index 2 and "A" is assigned the index 1 (even though that is not the order they appear in the data). These indices are then supplied to `igraph::contract()`, but the `mapping` argument has a different meaning; `mapping` is used to specify how nodes from the current graph should be mapped to nodes in the contracted graph. For example, for a 4-node graph `mapping = c(1, 1, 2, 2)` would mean that nodes 1 and 2 of the current graph are mapped to node 1 in the contracted graphs, and nodes 3 and 4 in the current graph are mapped to node 2 in the contracted graph.

In this example, `ind` has the value `c(2, 1)`, which is then interpreted by `igraph::contract()` as: map node 1 in the current graph to node 2 in the contracted graph, and map node 2 in the current graph to node 1 in the contracted graph. For the example, the net result is that the graph gets flipped.

In short: `group_indices()` and `mapping` have different meanings. It works out OK when `group_indices()` defines groups in the same order as they appear in the data, but breaks down when it orders the groups in a different way.

I could see this potentially causing issues whenever you contract a graph that has one or more communities consisting of a single node.

Also see tidyverse/dplyr#5668

_Originally posted by @mkoohafkan in https://github.com/thomasp85/tidygraph/issues/65#issuecomment-749636482_

**Comment by mkoohafkan** [comment]
I think the fix is to add this line immediately after the creation of `ind` (i.e., insert below line 251 of [tidygraph/R/morphers.R](https://github.com/thomasp85/tidygraph/blob/75c814ee9c45b2514396209939b7f06509a36b8a/R/morphers.R#L251)):
```r
ind <- match(ind, unique(ind))
```
This reorders the indices back to how they appear in the data, which (I think) returns a result consistent with how the `mapping` argument of `igraph::contract()` works. It seems to work with this simple example:
```r
nodes <- as_tibble(demo_netw, active = 'nodes')
nodes <- group_by(nodes, node_id)
ind <- group_indices(nodes)
ind <- match(ind, unique(ind))
ind
## [1] 1 2
```
